source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.h

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * Wrapper struct for gradient controller stats. @see stats_macros.h
+ */
+struct GradientControllerStats {
+  ALL_GRADIENT_CONTROLLER_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+class GradientControllerConfig {
+public:
+  GradientControllerConfig(
+      const envoy::config::filter::http::adaptive_concurrency::v2alpha::GradientControllerConfig&
+          proto_config);
+
+  // Accessors.
+  std::chrono::milliseconds min_rtt_calc_interval() const { return min_rtt_calc_interval_; }
+  std::chrono::milliseconds sample_rtt_calc_interval() const { return sample_rtt_calc_interval_; }
+  uint64_t max_concurrency_limit() const { return max_concurrency_limit_; }
+  int min_rtt_aggregate_request_count() const { return min_rtt_aggregate_request_count_; }
+  double max_gradient() const { return max_gradient_; }
+  double sample_aggregate_percentile() const { return sample_aggregate_percentile_; }
+
+private:
+  // The measured request round-trip time under ideal conditions.
+  std::chrono::milliseconds min_rtt_calc_interval_;
+
+  // The measured sample round-trip time from the previous time window.
+  std::chrono::milliseconds sample_rtt_calc_interval_;
+
+  // The maximum allowed concurrency value.
+  uint64_t max_concurrency_limit_;
+
+  // The number of requests to aggregate/sample during the minRTT recalculation.
+  int min_rtt_aggregate_request_count_;
+
+  // The maximum value the gradient may take.
+  double max_gradient_;
+
+  // The percentile value considered when processing samples.
+  double sample_aggregate_percentile_;
+};
+using GradientControllerConfigSharedPtr = std::shared_ptr<GradientControllerConfig>;
+
+/**
+ * A concurrency controller that implements the a variation of the Gradient algorithm to control the
+ * allowed concurrency window.
+ *
+ * An ideal round-trip time (minRTT) is measured periodically by only allowing a single outstanding
+ * request at a time and measuring the round-trip time to the listener. This information is then
+ * used in the calculation of a number called the gradient, using time-sampled latencies
+ * (sampleRTT):
+ *
+ *     gradient = minRTT / sampleRTT
+ *
+ * This gradient value has a useful property, such that it decreases as the sampled latencies
+ * increase. The value is then used to periodically update the concurrency limit via:
+ *
+ *     limit = old_limit * gradient
+ *     new_limit = limit + headroom
+ *
+ * The headroom value allows for request bursts and is also the driving factor behind increasing the
+ * concurrency limit when the sampleRTT is in the same ballpark as the minRTT. This value must be
+ * present in the calculation, since it forces the concurrency limit to increase until there is a
+ * deviation from the minRTT latency. In its absence, the concurrency limit could remain stagnant at
+ * an unnecessarily small value if sampleRTT ~= minRTT. Therefore, the headroom value is
+ * unconfigurable and is set to the square-root of the new limit.
```

Comment:
Presumably most/all of this is copied from the netflix code? Can you link back to relevant source material for bother further reading/history as well as credit?